File: app/src/main/java/in/woloo/www/common/DayLogUtils.java
package in.woloo.www.common;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import in.woloo.www.R;
import in.woloo.www.app.WolooApplication;
import in.woloo.www.dailylogscreen.models.DailyLogSubTitle;
import in.woloo.www.dailylogscreen.models.DailyLogSymptoms;
import in.woloo.www.database.preference.SharedPreference;
import in.woloo.www.database.preference.SharedPreferencesEnum;
import in.woloo.www.period_tracker.model.Log;

public class DayLogUtils {
    private final SharedPreference sharedPreference;

    private static DayLogUtils dayLogUtils;
    Map<String, Integer> preMenstruation = new LinkedHashMap<>();
    Map<String, Integer> menstruation = new LinkedHashMap<>();
    Map<String, Integer> diseaseAndMedication = new LinkedHashMap<>();
    Map<String, Integer> habits = new LinkedHashMap<>();

    Map<String, Integer> bleeding = new LinkedHashMap<>();
    Map<String, Integer> mood = new LinkedHashMap<>();
    Map<String, Integer> sexAndSexDrive = new LinkedHashMap<>();

    private DayLogUtils() {
        sharedPreference = new SharedPreference(WolooApplication.getInstance());
        preparePreMenstruationItems();
        prepareMenstruationItems();
        prepareDiseaseAndMedicationItems();
        prepareBleedingItems();
        prepareHabitsItems();
        prepareMoodItems();
        prepareSexAndSexDriveItems();
    }

    public static DayLogUtils getInstance() {
        if (dayLogUtils == null) {
            dayLogUtils = new DayLogUtils();
        }
        return dayLogUtils;
    }


    public List<DailyLogSymptoms> getDailyLog(SharedPreferencesEnum logType) {
        List<DailyLogSymptoms> dailyLogSymptoms = Collections.emptyList();
        switch (logType) {
            case BLEEDING:
                dailyLogSymptoms = getDailyLogInternal(SharedPreferencesEnum.BLEEDING);
                break;
            case MOOD:
                dailyLogSymptoms = getDailyLogInternal(SharedPreferencesEnum.MOOD);
                break;
            case SYMPTOMS:
                dailyLogSymptoms = getDailyLogInternal(SharedPreferencesEnum.SYMPTOMS);
                break;
            case DISEASES_AND_MEDICATION:
                dailyLogSymptoms = getDailyLogInternal(SharedPreferencesEnum.DISEASES_AND_MEDICATION);
                break;
            case SEX_AND_SEX_DRIVE:
                dailyLogSymptoms = getDailyLogInternal(SharedPreferencesEnum.SEX_AND_SEX_DRIVE);
                break;
            case HABITS:
                dailyLogSymptoms = getDailyLogInternal(SharedPreferencesEnum.HABITS);
                break;
            case PREMENSTRUATION:
                dailyLogSymptoms = getDailyLogInternal(SharedPreferencesEnum.PREMENSTRUATION);
                break;
            case MENSTRUATION:
                dailyLogSymptoms = getDailyLogInternal(SharedPreferencesEnum.MENSTRUATION);
                break;
        }
        return dailyLogSymptoms;
    }

    private List<DailyLogSymptoms> getDailyLogInternal(SharedPreferencesEnum logType) {
        String dailyLogStr = sharedPreference.getStoredPreference(WolooApplication.getInstance(), logType.getPreferenceKey());
        if (dailyLogStr == null) {
            return Collections.emptyList();
        }
        TypeToken<List<DailyLogSymptoms>> token = new TypeToken<List<DailyLogSymptoms>>() {
        };
        Gson gson = new Gson();
        return gson.fromJson(dailyLogStr, token.getType());
    }

    public List<DailyLogSubTitle> getCheckedDailyLog(SharedPreferencesEnum type) {
        List<DailyLogSymptoms> premenstrualLogAll = DayLogUtils.getInstance().getDailyLogAll(type);
        List<DailyLogSymptoms> premenstrualLog = DayLogUtils.getInstance().getDailyLog(type);
        return premenstrualLogAll.stream().map(dailyLog
                -> new DailyLogSubTitle(dailyLog.subTitleName, dailyLog.imageURL, premenstrualLog.contains(dailyLog))).collect(Collectors.toList());
    }

    public List<DailyLogSymptoms> getDailyLogAll(SharedPreferencesEnum logType) {
        switch (logType) {
            case BLEEDING:
                return setDailyLog(SharedPreferencesEnum.BLEEDING, new ArrayList<>(bleeding.keySet()), false);
            case MOOD:
                return setDailyLog(SharedPreferencesEnum.MOOD, new ArrayList<>(mood.keySet()), false);
            case DISEASES_AND_MEDICATION:
                return setDailyLog(SharedPreferencesEnum.DISEASES_AND_MEDICATION, new ArrayList<>(diseaseAndMedication.keySet()), false);
            case SEX_AND_SEX_DRIVE:
                return setDailyLog(SharedPreferencesEnum.SEX_AND_SEX_DRIVE, new ArrayList<>(sexAndSexDrive.keySet()), false);
            case HABITS:
                return setDailyLog(SharedPreferencesEnum.HABITS, new ArrayList<>(habits.keySet()), false);
            case PREMENSTRUATION:
                return setDailyLog(SharedPreferencesEnum.PREMENSTRUATION, new ArrayList<>(preMenstruation.keySet()), false);
            case MENSTRUATION:
                return setDailyLog(SharedPreferencesEnum.MENSTRUATION, new ArrayList<>(menstruation.keySet()), false);
        }
        return Collections.emptyList();
    }

    public Log getAsLog() {
        Log log = new Log();
        log.setBleeding(logStringValues(SharedPreferencesEnum.BLEEDING));
        log.setPremenstruation(logStringValues(SharedPreferencesEnum.PREMENSTRUATION));
        log.setMenstruation(logStringValues(SharedPreferencesEnum.MENSTRUATION));
        log.setMood(logStringValues(SharedPreferencesEnum.MOOD));
        log.setHabits(logStringValues(SharedPreferencesEnum.HABITS));
        log.setSexDrive(logStringValues(SharedPreferencesEnum.SEX_AND_SEX_DRIVE));
        log.setDiseasesandmedication(logStringValues(SharedPreferencesEnum.DISEASES_AND_MEDICATION));
        return log;
    }

    private List<String> logStringValues(SharedPreferencesEnum preferencesEnum) {
        List<DailyLogSymptoms> dailyLogs = getDailyLog(preferencesEnum);
        if (dailyLogs.size() > 0) {
            List<String> logValue = new ArrayList<>(dailyLogs.size());
            for (DailyLogSymptoms logSymptom : dailyLogs) {
                logValue.add(logSymptom.subTitleName);
            }
            return logValue;
        }
        return Collections.emptyList();
    }

    public void setDailyLog(SharedPreferencesEnum logType, List<String> dailyLogs) {
        setDailyLog(logType, dailyLogs, true);
    }

    public List<DailyLogSymptoms> setDailyLog(SharedPreferencesEnum logType, List<String> dailyLogs, boolean save) {
        List<DailyLogSymptoms> dailyLogSymptoms = new ArrayList<>();
        Gson gson = new Gson();
        for (String bleedingStr : dailyLogs) {
            switch (logType) {
                case PREMENSTRUATION:
                    if (preMenstruation.keySet().contains(bleedingStr)) {
                        DailyLogSymptoms dailyLogSymptom = prepareDailyLogEntry(bleedingStr, preMenstruation.get(bleedingStr));
                        dailyLogSymptoms.add(dailyLogSymptom);
                    }
                    break;
                case MENSTRUATION:
                    if (menstruation.keySet().contains(bleedingStr)) {
                        DailyLogSymptoms dailyLogSymptom = prepareDailyLogEntry(bleedingStr, menstruation.get(bleedingStr));
                        dailyLogSymptoms.add(dailyLogSymptom);
                    }
                    break;
                case BLEEDING:
                    if (bleeding.keySet().contains(bleedingStr)) {
                        DailyLogSymptoms dailyLogSymptom = prepareDailyLogEntry(bleedingStr, bleeding.get(bleedingStr));
                        dailyLogSymptoms.add(dailyLogSymptom);
                    }
                    break;
                case MOOD:
                    if (mood.keySet().contains(bleedingStr)) {
                        DailyLogSymptoms dailyLogSymptom = prepareDailyLogEntry(bleedingStr, mood.get(bleedingStr));
                        dailyLogSymptoms.add(dailyLogSymptom);
                    }
                    break;
                case DISEASES_AND_MEDICATION:
                    if (diseaseAndMedication.keySet().contains(bleedingStr)) {
                        DailyLogSymptoms dailyLogSymptom = prepareDailyLogEntry(bleedingStr, diseaseAndMedication.get(bleedingStr));
                        dailyLogSymptoms.add(dailyLogSymptom);
                    }
                    break;
                case HABITS:
                    if (habits.keySet().contains(bleedingStr)) {
                        DailyLogSymptoms dailyLogSymptom = prepareDailyLogEntry(bleedingStr, habits.get(bleedingStr));
                        dailyLogSymptoms.add(dailyLogSymptom);
                    }
                    break;
                case SEX_AND_SEX_DRIVE:
                    if (sexAndSexDrive.keySet().contains(bleedingStr)) {
                        DailyLogSymptoms dailyLogSymptom = prepareDailyLogEntry(bleedingStr, sexAndSexDrive.get(bleedingStr));
                        dailyLogSymptoms.add(dailyLogSymptom);
                    }
                    break;
            }
        }
        if (save) {
            String jsonString = gson.toJson(dailyLogSymptoms);
            sharedPreference.setStoredPreference(WolooApplication.getInstance(), logType.getPreferenceKey(), jsonString);
        }
        return dailyLogSymptoms;
    }

    private DailyLogSymptoms prepareDailyLogEntry(String bleedingStr, Integer integer) {
        DailyLogSymptoms dailyLogSymptoms = new DailyLogSymptoms();
        dailyLogSymptoms.setSubTitleName(bleedingStr);
        dailyLogSymptoms.setImageURL(integer);
        return dailyLogSymptoms;
    }

    private void preparePreMenstruationItems() {
        preMenstruation.put("Everything is fine", R.drawable.ic_symptoms_pre_mensturation_icon_one);
        preMenstruation.put("Cramps", R.drawable.ic_symptoms_pre_mensturation_icon_two);
        preMenstruation.put("Bloating in Lower Abdomen", R.drawable.ic_symptoms_pre_mensturation_icon_three);
        preMenstruation.put("Constipation", R.drawable.ic_symptoms_pre_mensturation_icon_four);
        preMenstruation.put("Heaviness in legs", R.drawable.ic_symptoms_pre_mensturation_icon_five);
        preMenstruation.put("Migrane", R.drawable.ic_symptoms_pre_mensturation_icon_six);
        preMenstruation.put("Change in appetite", R.drawable.ic_symptoms_pre_mensturation_icon_seven);
        preMenstruation.put("Headache", R.drawable.ic_symptoms_pre_mensturation_icon_eight);
    }

    private void prepareMenstruationItems() {
        menstruation.put("Everything is fine", R.drawable.ic_symptoms_mensturation_icon_one);
        menstruation.put("Tender Breasts", R.drawable.ic_symptoms_mensturation_icon_two);
        menstruation.put("Bowel", R.drawable.ic_symptoms_mensturation_icon_three);
        menstruation.put("Vomiting", R.drawable.ic_symptoms_mensturation_icon_four);
        menstruation.put("Headache", R.drawable.ic_symptoms_mensturation_icon_five);
        menstruation.put("Change in appetite", R.drawable.ic_symptoms_mensturation_icon_six);
    }

    private void prepareDiseaseAndMedicationItems() {
        diseaseAndMedication.put("Obesity", R.drawable.ic_diseases_medication_icon_one);
        diseaseAndMedication.put("Hypertension", R.drawable.ic_diseases_medication_icon_two);
        diseaseAndMedication.put("Diabetes", R.drawable.ic_diseases_medication_icon_three);
        diseaseAndMedication.put("PCOS", R.drawable.ic_diseases_medication_icon_four);
        diseaseAndMedication.put("Thyroid", R.drawable.ic_diseases_medication_icon_five);
        diseaseAndMedication.put("Stress", R.drawable.ic_diseases_medication_icon_six);
        diseaseAndMedication.put("Sickness", R.drawable.ic_diseases_medication_icon_seven);
        diseaseAndMedication.put("On Antibiotics", R.drawable.ic_diseases_medication_icon_eight);
        diseaseAndMedication.put("Any other medicine", R.drawable.ic_diseases_medication_icon_nine);
    }

    private void prepareHabitsItems() {
        habits.put("Smoking", R.drawable.ic_habits_icon_one);
        habits.put("drinking", R.drawable.ic_habits_icon_two);
    }

    private void prepareBleedingItems() {
        bleeding.put("Light", R.drawable.ic_bleeding_one_medium);
        bleeding.put("Medium", R.drawable.ic_bleeding_two_medium);
        bleeding.put("Heavy", R.drawable.ic_bleeding_four_heavy);
        bleeding.put("Spotting", R.drawable.ic_bleeding_one_spotting);
    }

    private void prepareMoodItems() {
        mood.put("Normal", R.drawable.ic_mood_icon_one_normal);
        mood.put("Happy", R.drawable.ic_mood_icon_two_happy);
        mood.put("Dizzy", R.drawable.ic_mood_icon_three_dizzy);
        mood.put("Tired", R.drawable.ic_mood_icon_three_tired);
    }

    private void prepareSexAndSexDriveItems() {
        sexAndSexDrive.put("Didn't have Sex", R.drawable.ic_sexdrive_one);
        sexAndSexDrive.put("Protected Sex", R.drawable.ic_sexdrive_two);
        sexAndSexDrive.put("Unprotected Sex", R.drawable.ic_sexdrive_three);
        sexAndSexDrive.put("High Sex Drive", R.drawable.ic_sexdrive_four);
        sexAndSexDrive.put("masturbation", R.drawable.ic_sexdrive_five);
    }
}

File: app/src/main/java/in/woloo/www/application_kotlin/utilities/DayLogUtils.kt
package `in`.woloo.www.application_kotlin.utilities

import android.content.res.ColorStateList
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import `in`.woloo.www.R
import `in`.woloo.www.app.WolooApplication
import `in`.woloo.www.more.dailylogscreen.models.DailyLogSubTitle
import `in`.woloo.www.more.dailylogscreen.models.DailyLogSymptoms
import `in`.woloo.www.application_kotlin.database.SharedPreference
import `in`.woloo.www.database.preference.SharedPreferencesEnum
import `in`.woloo.www.more.period_tracker.model.Log
import java.util.function.Function
import java.util.stream.Collectors

class DayLogUtils private constructor() {
    private val sharedPreference: SharedPreference = SharedPreference(WolooApplication.instance!!)
    private var preMenstruation: MutableMap<String, Int> = LinkedHashMap()
    var menstruation: MutableMap<String, Int> = LinkedHashMap()
    private var diseaseAndMedication: MutableMap<String, Int> = LinkedHashMap()
    private var habits: MutableMap<String, Int> = LinkedHashMap()
    var bleeding: MutableMap<String, Int> = LinkedHashMap()
    var mood: MutableMap<String, Int> = LinkedHashMap()
    private var sexAndSexDrive: MutableMap<String, Int> = LinkedHashMap()



    init {
        preparePreMenstruationItems()
        prepareMenstruationItems()
        prepareDiseaseAndMedicationItems()
        prepareBleedingItems()
        prepareHabitsItems()
        prepareMoodItems()
        prepareSexAndSexDriveItems()
    }

    fun getDailyLog(logType: SharedPreferencesEnum?): List<DailyLogSymptoms?> {
        var dailyLogSymptoms: List<DailyLogSymptoms?> = emptyList<DailyLogSymptoms>()
        when (logType) {
            SharedPreferencesEnum.BLEEDING -> dailyLogSymptoms =
                getDailyLogInternal(SharedPreferencesEnum.BLEEDING)

            SharedPreferencesEnum.MOOD -> dailyLogSymptoms =
                getDailyLogInternal(SharedPreferencesEnum.MOOD)

            SharedPreferencesEnum.SYMPTOMS -> dailyLogSymptoms =
                getDailyLogInternal(SharedPreferencesEnum.SYMPTOMS)

            SharedPreferencesEnum.DISEASES_AND_MEDICATION -> dailyLogSymptoms =
                getDailyLogInternal(SharedPreferencesEnum.DISEASES_AND_MEDICATION)

            SharedPreferencesEnum.SEX_AND_SEX_DRIVE -> dailyLogSymptoms =
                getDailyLogInternal(SharedPreferencesEnum.SEX_AND_SEX_DRIVE)

            SharedPreferencesEnum.HABITS -> dailyLogSymptoms =
                getDailyLogInternal(SharedPreferencesEnum.HABITS)

            SharedPreferencesEnum.PREMENSTRUATION -> dailyLogSymptoms =
                getDailyLogInternal(SharedPreferencesEnum.PREMENSTRUATION)

            SharedPreferencesEnum.MENSTRUATION -> dailyLogSymptoms =
                getDailyLogInternal(SharedPreferencesEnum.MENSTRUATION)

            else -> {}
        }
        return dailyLogSymptoms
    }

    private fun getDailyLogInternal(logType: SharedPreferencesEnum): List<DailyLogSymptoms?> {
        val dailyLogStr: String = sharedPreference.getStoredPreference(
            WolooApplication.instance!!,
            logType.getPreferenceKey()
        )
            ?: return emptyList<DailyLogSymptoms>()
        val token: TypeToken<List<DailyLogSymptoms?>?> =
            object : TypeToken<List<DailyLogSymptoms?>?>() {}
        val gson = Gson()
        return gson.fromJson<List<DailyLogSymptoms?>>(dailyLogStr, token.type)
    }

    fun getCheckedDailyLog(type: SharedPreferencesEnum?): List<DailyLogSubTitle> {
        val premenstrualLogAll: List<DailyLogSymptoms> = instance!!.getDailyLogAll(type)
        val premenstrualLog: List<DailyLogSymptoms?> = instance!!.getDailyLog(type)
        return premenstrualLogAll.stream()
            .map<DailyLogSubTitle>(Function<DailyLogSymptoms, DailyLogSubTitle> { dailyLog: DailyLogSymptoms ->
                DailyLogSubTitle(
                    dailyLog.subTitleName,
                    dailyLog.imageURL,
                    premenstrualLog.contains(dailyLog)
                )
            }).collect(
            Collectors.toList<DailyLogSubTitle>()
        )
    }

    fun getDailyLogAll(logType: SharedPreferencesEnum?): List<DailyLogSymptoms> {
        when (logType) {
            SharedPreferencesEnum.BLEEDING ->
                return setDailyLog(
                SharedPreferencesEnum.BLEEDING,
                ArrayList<String>(bleeding.keys),
                false
            )

            SharedPreferencesEnum.MOOD -> return setDailyLog(
                SharedPreferencesEnum.MOOD,
                ArrayList<String>(mood.keys),
               false
            )

            SharedPreferencesEnum.DISEASES_AND_MEDICATION -> return setDailyLog(
                SharedPreferencesEnum.DISEASES_AND_MEDICATION,
                ArrayList<String>(diseaseAndMedication.keys),
                false
            )

            SharedPreferencesEnum.SEX_AND_SEX_DRIVE -> return setDailyLog(
                SharedPreferencesEnum.SEX_AND_SEX_DRIVE,
                ArrayList<String>(sexAndSexDrive.keys),
              false
            )

            SharedPreferencesEnum.HABITS -> return setDailyLog(
                SharedPreferencesEnum.HABITS,
                ArrayList<String>(habits.keys),
               false
            )

            SharedPreferencesEnum.PREMENSTRUATION -> return setDailyLog(
                SharedPreferencesEnum.PREMENSTRUATION,
                ArrayList<String>(preMenstruation.keys),
                false
            )

            SharedPreferencesEnum.MENSTRUATION -> return setDailyLog(
                SharedPreferencesEnum.MENSTRUATION,
                ArrayList<String>(menstruation.keys),
              false
            )

            else -> {}
        }
        return emptyList<DailyLogSymptoms>()
    }

    val asLog: Log
        get() {
            val log = Log()
            log.bleeding = logStringValues(SharedPreferencesEnum.BLEEDING)
            log.premenstruation = logStringValues(SharedPreferencesEnum.PREMENSTRUATION)
            log.menstruation = logStringValues(SharedPreferencesEnum.MENSTRUATION)
            log.mood = logStringValues(SharedPreferencesEnum.MOOD)
            log.habits = logStringValues(SharedPreferencesEnum.HABITS)
            log.sexDrive = logStringValues(SharedPreferencesEnum.SEX_AND_SEX_DRIVE)
            log.diseasesandmedication =
                logStringValues(SharedPreferencesEnum.DISEASES_AND_MEDICATION)
            return log
        }

    private fun logStringValues(preferencesEnum: SharedPreferencesEnum): List<String> {
        val dailyLogs: List<DailyLogSymptoms?> = getDailyLog(preferencesEnum)
        if (dailyLogs.size > 0) {
            val logValue: MutableList<String> = ArrayList(dailyLogs.size)
            for (logSymptom in dailyLogs) {
                logValue.add(logSymptom!!.subTitleName!!)
            }
            return logValue
        }
        return emptyList()
    }

    fun setDailyLog(logType: SharedPreferencesEnum, dailyLogs: List<String> ) {
        setDailyLog(logType, dailyLogs, true)
    }

    fun setDailyLog(
        logType: SharedPreferencesEnum,
        dailyLogs: List<String>,
        save: Boolean
    ): List<DailyLogSymptoms> {
        val dailyLogSymptoms: MutableList<DailyLogSymptoms> = ArrayList<DailyLogSymptoms>()
        val gson = Gson()
        for (bleedingStr in dailyLogs) {
            when (logType) {
                SharedPreferencesEnum.PREMENSTRUATION -> if (preMenstruation.keys.contains(
                        bleedingStr
                    )
                ) {
                    val dailyLogSymptom: DailyLogSymptoms =
                        prepareDailyLogEntry(bleedingStr, preMenstruation[bleedingStr])
                    dailyLogSymptoms.add(dailyLogSymptom)
                }

                SharedPreferencesEnum.MENSTRUATION -> if (menstruation.keys.contains(bleedingStr)) {
                    val dailyLogSymptom: DailyLogSymptoms =
                        prepareDailyLogEntry(bleedingStr, menstruation[bleedingStr])
                    dailyLogSymptoms.add(dailyLogSymptom)
                }

                SharedPreferencesEnum.BLEEDING -> if (bleeding.keys.contains(bleedingStr)) {
                    val dailyLogSymptom: DailyLogSymptoms =
                        prepareDailyLogEntry(bleedingStr, bleeding[bleedingStr])
                    dailyLogSymptoms.add(dailyLogSymptom)
                }

                SharedPreferencesEnum.MOOD -> if (mood.keys.contains(bleedingStr)) {
                    val dailyLogSymptom: DailyLogSymptoms =
                        prepareDailyLogEntry(bleedingStr, mood[bleedingStr])
                    dailyLogSymptoms.add(dailyLogSymptom)
                }

                SharedPreferencesEnum.DISEASES_AND_MEDICATION -> if (diseaseAndMedication.keys.contains(
                        bleedingStr
                    )
                ) {
                    val dailyLogSymptom: DailyLogSymptoms =
                        prepareDailyLogEntry(bleedingStr, diseaseAndMedication[bleedingStr])
                    dailyLogSymptoms.add(dailyLogSymptom)
                }

                SharedPreferencesEnum.HABITS -> if (habits.keys.contains(bleedingStr)) {
                    val dailyLogSymptom: DailyLogSymptoms =
                        prepareDailyLogEntry(bleedingStr, habits[bleedingStr])
                    dailyLogSymptoms.add(dailyLogSymptom)
                }

                SharedPreferencesEnum.SEX_AND_SEX_DRIVE -> if (sexAndSexDrive.keys.contains(
                        bleedingStr
                    )
                ) {
                    val dailyLogSymptom: DailyLogSymptoms =
                        prepareDailyLogEntry(bleedingStr, sexAndSexDrive[bleedingStr])
                    dailyLogSymptoms.add(dailyLogSymptom)
                }

                else -> {}
            }
        }
        if (save) {
            val jsonString = gson.toJson(dailyLogSymptoms)
            sharedPreference.setStoredPreference(
                WolooApplication.instance!!,
                logType.preferenceKey,
                jsonString
            )
        }
        return dailyLogSymptoms
    }

    private fun prepareDailyLogEntry(bleedingStr: String, integer: Int?): DailyLogSymptoms {
        val dailyLogSymptoms = DailyLogSymptoms()
        dailyLogSymptoms.subTitleName = bleedingStr
        if (integer != null) {
            dailyLogSymptoms.imageURL = integer
        }
        return dailyLogSymptoms
    }

    private fun preparePreMenstruationItems() {
        preMenstruation["Everything is fine"] = R.drawable.all_like
        preMenstruation["Cramps"] = R.drawable.cramps_icon
        preMenstruation["Bloating in Lower Abdomen"] =
            R.drawable.bloating_icon
        preMenstruation["Constipation"] = R.drawable.constipation_icon
        preMenstruation["Heaviness in legs"] = R.drawable.cramp_legs_icon
        preMenstruation["Migrane"] = R.drawable.migraine_icon
        preMenstruation["Change in appetite"] = R.drawable.loss_appetite_icon
        preMenstruation["Headache"] = R.drawable.pain_icon
    }

    private fun prepareMenstruationItems() {
        menstruation["Everything is fine"] = R.drawable.all_like
        menstruation["Tender Breasts"] = R.drawable.tender_breast_icon
        menstruation["Bowel"] = R.drawable.bowel_icon
        menstruation["Vomiting"] = R.drawable.vomiting_icon
        menstruation["Headache"] = R.drawable.pain_icon
        menstruation["Change in appetite"] = R.drawable.loss_appetite_icon
    }

    private fun prepareDiseaseAndMedicationItems() {
        diseaseAndMedication["Obesity"] = R.drawable.obesity_icon_img
        diseaseAndMedication["Hypertension"] = R.drawable.hypertension_icon_img
        diseaseAndMedication["Diabetes"] = R.drawable.diabetes_icon_img
        diseaseAndMedication["PCOS"] = R.drawable.pcod_icon_img
        diseaseAndMedication["Thyroid"] = R.drawable.thyroid_icon_img
        diseaseAndMedication["Stress"] = R.drawable.stress_icon_img
        diseaseAndMedication["Sickness"] = R.drawable.sickness_icon
        diseaseAndMedication["On Antibiotics"] = R.drawable.antibiotic_icon
        diseaseAndMedication["Any other medicine"] = R.drawable.medicine_icon
    }

    private fun prepareHabitsItems() {
        habits["Smoking"] = R.drawable.cigarette_icon
        habits["drinking"] = R.drawable.liquor_icon
    }

    private fun prepareBleedingItems() {
        bleeding["Light"] = R.drawable.light_periods_icon
        bleeding["Medium"] = R.drawable.medium_periods_icon
        bleeding["Heavy"] = R.drawable.heavy_period_icons
        bleeding["Spotting"] = R.drawable.spotting_period_icons
    }

    private fun prepareMoodItems() {
        mood["Normal"] = R.drawable.normal_mood_iocon
        mood["Happy"] = R.drawable.happy_mood_icon
        mood["Dizzy"] = R.drawable.dizzy_mood_icon
        mood["Tired"] = R.drawable.tired_mood_icon
    }

    private fun prepareSexAndSexDriveItems() {
        sexAndSexDrive["Didn't have Sex"] = R.drawable.no_sex_icon
        sexAndSexDrive["Protected Sex"] = R.drawable.protected_sex_icon
        sexAndSexDrive["Unprotected Sex"] = R.drawable.unprotected_sex_icon
        sexAndSexDrive["High Sex Drive"] = R.drawable.high_sex_icon
        sexAndSexDrive["masturbation"] = R.drawable.masturbation_icon
    }

    companion object {
        private var dayLogUtils: DayLogUtils? = null
        @JvmStatic
        val instance: DayLogUtils?
            get() {
                if (dayLogUtils == null) {
                    dayLogUtils = DayLogUtils()
                }
                return dayLogUtils
            }
    }
}
